### PR TITLE
Add prometheus exporter

### PR DESCRIPTION
**Pull request by bernd**
This implements a prometheus exporter that maps the internal dropwizard
metrics to prometheus compatible metric names and labels.

The metric mapping is configurable via config files.

Fixes https://github.com/Graylog2/graylog2-server/issues/9914

## TODO

- [ ] Add some tests
- [ ] Register our default exporter port in the [prometheus default port registry](https://github.com/prometheus/prometheus/wiki/Default-port-allocations)
- [ ] Check for TODOs and commented code
- [x] Check that the metrics filter works correctly
- [x] Add error handling and constant for the built-in resource file
- [ ] Add HTTP endpoint that returns a JSON payload for the [file based discovery](https://prometheus.io/docs/guides/file-sd/) in prometheus
- [ ] Test with a real prometheus

### CODE REVIEW

File: misc/graylog.conf [hunk 1]
```diff
@@ -689,3 +689,27 @@ proxied_requests_thread_pool_size = 32
...
+
+# IP address and port for the prometheus exporter HTTP server.
+# Default: 127.0.0.1:9832
+#prometheus_exporter_bind_address = 127.0.0.1:9832
+
+# Path to the prometheus exporter core mapping file. If this option is enabled, the full built-in core mapping is
+# replaced with the mappings in this file.
+# This file is monitored for changes and updates will be applied at runtime.
+# Default: ./prometheus-exporter-mapping-core.yml
+#prometheus_exporter_mapping_file_path_core = prometheus-exporter-mapping-core.yml
```
Comment (danotorrey) [inline misc/graylog.conf:705]: Should the core file mappings really be configurable? As far as I can see, there is only benefit in allowing the custom optional mappings to be configured.

Perhaps storing the core mappings as a compiled resource is enough. What do you think?

**Review by danotorrey — COMMENTED** [review]
@bernd I really like the approach and design here. It's a nice improvement, which allows for configurability, and improves the definition of the mappings (both core and custom). I really did not like how the mappings definitions were explicitly defined  in Java code in the first implementation. YML works very well as far as I can see here. 

I still need to do more review and testing, but it does appear to work. I will take one more pass through  with deeper testing. I left a few questions from the first look.

File: misc/graylog.conf [hunk 2]
```diff
@@ -689,3 +689,27 @@ proxied_requests_thread_pool_size = 32
 # Setting this to an empty value, leaves it up to system libraries and the used JDK to chose a default.
 # Default: TLSv1.2,TLSv1.3  (might be automatically adjusted to protocols supported by the JDK)
 #enabled_tls_protocols= TLSv1.2,TLSv1.3
+
+# Enable prometheus exporter HTTP server.
```
[the remark this answers is not in the document]
Comment (bernd) [reply misc/graylog.conf]: Yes, good point. I fixed it.

File: pom.xml [hunk 3]
```diff
@@ -143,6 +143,7 @@
         <opencsv.version>2.3</opencsv.version>
         <os-platform-finder.version>1.2.3</os-platform-finder.version>
         <pkts.version>3.0.5</pkts.version>
+        <prometheus-client.version>0.10.0</prometheus-client.version>
```
[the remark this answers is not in the document]
Comment (bernd) [reply pom.xml]: Heh, yes. Updated!

File: misc/graylog.conf (hunk 1 above)
Comment (bernd) [reply misc/graylog.conf:705]: @danotorrey The idea was, that it's possible to change the core mapping without any code changes if really needed. (e.g. a bug fix)

This option is not something that is supposed to be used very often. We might want to mention that in the comment for the option?

Comment (danotorrey) [reply misc/graylog.conf:705]: As discussed, keeping the overridable core mappings makes sense.

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            LOG.debug("Detecting changes for file <{}>", file);
+
+            if (fileInfoRef.get() == null) {
+                LOG.debug("Getting initial file info for file <{}>", file);
+                fileInfoRef.set(FileInfo.forPath(file));
+            }
+
+            final FileInfo.Change change = fileInfoRef.get().checkForChange();
+
+            if (change.isChanged()) {
```
Comment (danotorrey) [inline graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java:79]: I think this check missing the edge case that the mappings (core or custom) file does not exist at server start time, then is added later.  Once the service starts, then the file is added later, the mappings are not loaded when the initial file info is loaded during refresh, so second change to the file is needed to force the change to be detected and the metrics in the file to be loaded.

**Review by danotorrey — CHANGES_REQUESTED** [review]
Well done! I tested this with a Prometheus instance running, and it works. The live refresh of the core/custom mappings is really cool. I am sure users will love that. 

From testing, I found a few items that I gave some feedback on. Once this is ready for review, I will do one last final pass through the code.

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+@SuppressWarnings({"FieldMayBeFinal", "unused"})
+public class PrometheusExporterConfiguration {
+    private static final String PREFIX = "prometheus_exporter_";
+
+    public static final String ENABLED = PREFIX + "enabled";
+    public static final String BIND_ADDRESS = "bind_address";
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java:34]: Good catch, thanks!

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    @Parameter(value = MAPPING_FILE_PATH_CORE, validators = {FilePathReadableValidator.class})
+    private Path mappingFilePathCore = Paths.get("prometheus-exporter-mapping-core.yml");
+
+    @Parameter(value = MAPPING_FILE_PATH_CUSTOM, validators = {FilePathReadableValidator.class})
+    private Path mappingFilePathCustom = Paths.get("prometheus-exporter-mapping-custom.yml");
+
+    @Parameter(value = MAPPING_FILE_REFRESH_INTERVAL, validators = {PositiveDurationValidator.class})
+    private Duration mappingFileRefreshInterval = Duration.seconds(60);
+
+    private static class CustomHostAndPortConverter extends HostAndPortConverter {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java:61]: Good catch, thanks!

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingConfigLoader.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    private final NodeId nodeId;
+
+    @Inject
+    public PrometheusMappingConfigLoader(NodeId nodeId) {
+        this.nodeId = nodeId;
+    }
+
+
+    public Set<MapperConfig> load(byte[] bytes) throws IOException {
+        final PrometheusMappingConfig config = YAML_MAPPER.readValue(new ByteArrayInputStream(bytes), PrometheusMappingConfig.class);
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingConfigLoader.java:44]: The error log message indicates the problematic file already. Do you suggest to remove the stack trace? Or what do you mean with a nicer error message? :slightly_smiling_face: 
```
2021-06-03 12:48:52,837 ERROR: org.graylog.metrics.prometheus.mapping.PrometheusMappingFilesHandler - Couldn't load mapping from file </path/to/file/prometheus-exporter-dev.yml>
com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.util.ArrayList` out of VALUE_STRING token
 at [Source: (ByteArrayInputStream); line: 4, column: 3] (through reference chain: org.graylog.metrics.prometheus.mapping.PrometheusMappingConfig["metric_mappings"])
```

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    public List<MapperConfig> getMapperConfigs() {
+        final Set<MapperConfig> mapperConfigs = new HashSet<>();
+
+        // If an external core mapping file exists it takes precedence over the included mapping resource
+        if (Files.exists(coreMappingFile)) {
+            LOG.debug("Loading core metric mappings from file <{}>", coreMappingFile);
+            try {
+                mapperConfigs.addAll(mapperConfigLoader.load(Files.readAllBytes(coreMappingFile)));
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java:115]: Good idea. I don't expect super large YAML files, but you never know. :smile:

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    private boolean enabled = false;
+
+    @Parameter(value = BIND_ADDRESS, converter = CustomHostAndPortConverter.class)
+    private HostAndPort bindAddress = HostAndPort.fromParts(DEFAULT_BIND_ADDRESS_HOST, DEFAULT_BIND_ADDRESS_PORT);
+
+    @Parameter(value = MAPPING_FILE_PATH_CORE, validators = {FilePathReadableValidator.class})
+    private Path mappingFilePathCore = Paths.get("prometheus-exporter-mapping-core.yml");
+
+    @Parameter(value = MAPPING_FILE_PATH_CUSTOM, validators = {FilePathReadableValidator.class})
+    private Path mappingFilePathCustom = Paths.get("prometheus-exporter-mapping-custom.yml");
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java:56]: @danotorrey I used a default to avoid null checks in the source code. But I think you are right, we shouldn't set a default in the code. This will fail with the config validator if the files don't exist. :smile:  We can always add some good default values to the config file for the OS packages so users only have to uncomment the entries. I will update the code.

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java (hunk 4 above)
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java:79]: Good catch, thanks! Fixed in https://github.com/Graylog2/graylog2-server/pull/10767/commits/fb8be5536c2814fb7ddb304ee7bd0ea1347d3c41.

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingFilesHandler.java]: Done in https://github.com/Graylog2/graylog2-server/pull/10767/commits/b81164c78b1fddc10f612039a5d7b4128f9ba923.

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/PrometheusExporterConfiguration.java]: Done in https://github.com/Graylog2/graylog2-server/pull/10767/commits/3f924c012c85f6b36c006e7dfb41e0ffae0d61ca.

File: graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingConfigLoader.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (danotorrey) [reply graylog2-server/src/main/java/org/graylog/metrics/prometheus/mapping/PrometheusMappingConfigLoader.java]: Oops, I missed that. Looks good to me as-is.

**Review by danotorrey — APPROVED** [review]
Looks good to me! Thanks a lot for the fixes. All tests successfully. I am merging this now.

I noticed one interesting aspect worth mentioning. Now that the defaults are not specified for the file, the edge case I mentioned is probably extremely hard to hit. Without defaults, the `FilePathReadableValidator.class` config validator seems to prevent nonexistent files from being specified on startup - effectively preventing the problematic situation. I think the defaults were exempt from the validation :D It's fine as is now since there's extra protection around it.

File: graylog2-server/src/main/resources/prometheus-exporter.yml [hunk 10]
```diff
@@ -0,0 +1,1344 @@
...
+  - metric_name: "auth_actions"
+    match_pattern: "org.graylog.plugins.cloud.auth.*.failure.gl.*"
+    wildcard_extract_labels:
+      - "action"
+      - "graylog_error_code"
+    additional_labels:
+      success: "false"
+      type: "graylog"
+
+  - metric_name: "input-buffer-thread-factory"
```
Comment (fungusakafungus) [inline graylog2-server/src/main/resources/prometheus-exporter.yml:1246]: Shouldn't these be underscores instead?

Comment (danotorrey) [reply graylog2-server/src/main/resources/prometheus-exporter.yml:1246]: @fungusakafungus You are absolutely right. Thank you very much for seeing this. https://github.com/Graylog2/graylog2-server/issues/10904 has been created, and we will fix this.

Comment (danotorrey) [reply graylog2-server/src/main/resources/prometheus-exporter.yml:1246]: This has been fixed in https://github.com/Graylog2/graylog2-server/pull/10905 and will be included in the 4.1 release.
